lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -187,6 +187,387 @@ static SmallVector<int64_t> computeDimsOrderForMoveDim(int64_t srcDimInt,
+                                         SmallVector<char> &rhsReduceDims,
+                                         SmallVector<char> &resultTokens) {
+  // generate ideal result dims, i.e.,
+  // [*batchingDims, *lhsOtherDims, *lhsReduceDims, *rhsOtherDims,
+  // *rhsReduceDims]
+  resultTokens.insert(resultTokens.end(), batchingDims.begin(),
+                      batchingDims.end());
+  resultTokens.insert(resultTokens.end(), lhsOtherDims.begin(),
+                      lhsOtherDims.end());
+  resultTokens.insert(resultTokens.end(), lhsReduceDims.begin(),
+                      lhsReduceDims.end());
+  resultTokens.insert(resultTokens.end(), rhsOtherDims.begin(),
+                      rhsOtherDims.end());
+  resultTokens.insert(resultTokens.end(), rhsReduceDims.begin(),
+                      rhsReduceDims.end());
+}
+
+static Value permuteTensorForMatmul(PatternRewriter &rewriter, Location loc,
+                                    Value input, SmallVector<char> &dimTokens,
+                                    SmallVector<char> &batchingDims,
+                                    SmallVector<char> &contractingDims,
+                                    SmallVector<char> &otherDims,
+                                    SmallVector<char> &reduceDims, bool isLhs) {
+  auto inputType = input.getType().cast<BaseTensorType>();
+  llvm::SmallDenseMap<char, int64_t> dimTokenMap;
+  for (size_t idx = 0; idx < dimTokens.size(); ++idx) {
+    dimTokenMap[dimTokens[idx]] = idx;
+  }
+
+  SmallVector<Value> permuteVec;
+  auto appendDims = [&](SmallVector<char> dimTokens) {
+    for (auto d : dimTokens) {
+      permuteVec.push_back(rewriter.create<Torch::ConstantIntOp>(
+          loc, rewriter.getI64IntegerAttr(dimTokenMap[d])));
+    }
+  };
+
+  appendDims(batchingDims);
+  if (!isLhs)
+    appendDims(contractingDims);
+  appendDims(otherDims);
+  appendDims(reduceDims);
+  if (isLhs)
+    appendDims(contractingDims);
+
+  Value dstDims = rewriter.create<Torch::PrimListConstructOp>(
+      loc, Torch::ListType::get(Torch::IntType::get(rewriter.getContext())),
+      permuteVec);
+  auto outType = inputType.getWithSizesAndDtype(std::nullopt,
+                                                inputType.getOptionalDtype());
+  return rewriter.create<Torch::AtenPermuteOp>(loc, outType, input, dstDims);
+}
+
+static LogicalResult performMatmul(PatternRewriter &rewriter, Location loc,
+                                   Value lhs, SmallVector<char> &lhsTokens,
+                                   Value rhs, SmallVector<char> &rhsTokens,
+                                   Value &result,
+                                   SmallVector<char> &resultTokens,
+                                   SmallVector<char> &finalResultTokens) {
+  auto lhsType = lhs.getType().cast<BaseTensorType>();
+  auto rhsType = rhs.getType().cast<BaseTensorType>();
+  Type promotedDType;
+
+  // promote dtype
+  if (lhsType.hasDtype() && rhsType.hasDtype()) {
```

Comment:
The type promotion should be handled by the dtype functions. What's the reason for not having shape and dtype functions?

Reply:
the problem here is that we have some cases that einsum operands can have differing dtypes, it becomes an issue when we attempt to decompose it into mm or bmm operations, dtype mismatches would result in errors. Also, shape inference for the einsum operation, which necessitates the analysis of the equation string that specifies the subscripts for summation, but right now it's not supported by abstract_interp_lib_gen

Reply:
> The type promotion should be handled by the dtype functions. What's the reason for not having shape and dtype functions?

the problem here is that we have some cases that einsum operands can have differing dtypes, it becomes an issue when we attempt to decompose it into mm or bmm operations, dtype mismatches would result in errors. Also, shape inference for the einsum operation, which necessitates the analysis of the equation string that specifies the subscripts for summation, but right now it's not supported by abstract_interp_lib_gen

Reply:
I would've imagined that the result type is simply the promotion of all the dtypes in the input tensors list. An example of how to handle a list of tensors is

https://github.com/llvm/torch-mlir/blob/53fc995639e26f40465516dded26548928bd2d1a/projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py#L3651

Also, let's add an e2e test that has inputs of different dtypes to test this behavior.

For the shape information, I think if we just leave the result types of the different `mm` and `bmm` ops as having no shape information, the shape propagation passes should take care of filling those in.

Reply:
> I would've imagined that the result type is simply the promotion of all the dtypes in the input tensors list. An example of how to handle a list of tensors is
> 
> https://github.com/llvm/torch-mlir/blob/53fc995639e26f40465516dded26548928bd2d1a/projects/pt1/python/torch_mlir/jit_ir_importer/build_tools/abstract_interp_lib_gen.py#L3651
> 
> Also, let's add an e2e test that has inputs of different dtypes to test this behavior.
> 
> For the shape information, I think if we just leave the result types of the different `mm` and `bmm` ops as having no shape information, the shape propagation passes should take care of filling those in.

I have removed the original type promotion part and added the corresponding type function. Additionally, I found that the Torch frontend does not allow for different input types in einsum, so it might be an issue arising during our model conversion process.